google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -3587,6 +3678,10 @@ public void onSessionReady(SessionImpl sessionImpl) {
     public void onSessionCreateFailure(Throwable t, int createFailureForSessionCount) {
       synchronized (lock) {
         multiplexedSessionBeingCreated = false;
+        if (isDatabaseOrInstanceNotFound(newSpannerException(t))) {
+          setResourceNotFoundException((ResourceNotFoundException) t);
+          poolMaintainer.close();
+        }
         currentMultiplexedSessionReference.get().setException(newSpannerException(t));
```

Comment:
@olavloite I think the order here does not matter. i.e it is safe to first do `poolMaintainer.close();` and then propagate the exceptions to all the waiting threads.

Reply:
Yes, I think so too.